File: solutions/2251. Number of Flowers in Full Bloom/2251.py
# Time complexity: O((n+m)logn)
# Space complexity: O(n+m)
class Solution:
  def fullBloomFlowers(
    self, flowers: List[List[int]], people: List[int]
  ) -> List[int]:
    starts = sorted(s for s, _ in flowers)
    ends = sorted(e for _, e in flowers)
    return [bisect.bisect_right(starts, person) -
            bisect.bisect_left(ends, person)
            for person in people]

File: solutions/2251. Number of Flowers in Full Bloom/2251.cpp
// Time complexity: O((n+m)logn)
// Space complexity: O(n+m)
class Solution {
 public:
  vector<int> fullBloomFlowers(
    vector<vector<int>>& flowers, vector<int>& people
  ) {
    vector<int> ans;
    vector<int> starts;
    vector<int> ends;

    for (const vector<int>& flower : flowers) {
      starts.push_back(flower[0]);
      ends.push_back(flower[1]);
    }

    ranges::sort(starts);
    ranges::sort(ends);

    for (const int p : people) {
      const int started = ranges::upper_bound(starts, p) - starts.begin();
      const int ended = ranges::lower_bound(ends, p) - ends.begin();
      ans.push_back(started - ended);
    }

    return ans;
  }
};